targets/TARGET_GigaDevice/TARGET_GD32F30X/i2c_api.c

```diff
@@ -0,0 +1,653 @@
+
+/** Reset I2C peripheral by hardware method. Most of the implementation enable RCU reset.
+ *
+ *  @param obj  The I2C object
+ */
+static void i2c_hw_reset(i2c_t *obj)
+{
+    struct i2c_s *obj_s = I2C_S(obj);
+
+    switch (obj_s->i2c) {
+        case I2C_0:
+            rcu_periph_reset_enable(RCU_I2C0RST);
+            rcu_periph_reset_disable(RCU_I2C0RST);
+            break;
+
+        case I2C_1:
+            rcu_periph_reset_enable(RCU_I2C1RST);
+            rcu_periph_reset_disable(RCU_I2C1RST);
+            break;
+    }
+}
+
+/** Initialize the I2C peripheral. It sets the default parameters for I2C
+ *  peripheral, and configures its specifieds pins.
+ *
+ *  @param obj  The I2C object
+ *  @param sda  The sda pin
+ *  @param scl  The scl pin
+ */
+void i2c_init(i2c_t *obj, PinName sda, PinName scl)
+{
+    struct i2c_s *obj_s = I2C_S(obj);
+
+    /* find the I2C by pins */
+    uint32_t i2c_sda = pinmap_peripheral(sda, PinMap_I2C_SDA);
+    uint32_t i2c_scl = pinmap_peripheral(scl, PinMap_I2C_SCL);
+
+    obj_s->sda = sda;
+    obj_s->scl = scl;
+    obj_s->i2c = (I2CName)pinmap_merge(i2c_sda, i2c_scl);
+    MBED_ASSERT(obj_s->i2c != (I2CName)NC);
+
+    switch (obj_s->i2c) {
+        case I2C_0:
+            /* enable I2C0 clock and configure the pins of I2C0 */
+            obj_s->index = 0;
+            rcu_periph_clock_enable(RCU_I2C0);
+
+            break;
+
+        case I2C_1:
+            /* enable I2C1 clock and configure the pins of I2C1 */
+            obj_s->index = 1;
+            rcu_periph_clock_enable(RCU_I2C1);
+
+            break;
+
+        default:
+            break;
+    }
+
+    /* configure the pins of I2C */
+    pinmap_pinout(sda, PinMap_I2C_SDA);
+    pinmap_pinout(scl, PinMap_I2C_SCL);
+
+    /* 100 KHz as the default I2C frequence */
+    i2c_frequency(obj, 100000);
+
+    obj_s->state = (operation_state_enum)I2C_STATE_NONE;
+    obj_s->previous_state_mode = I2C_STATE_NONE;
+    obj_s->global_trans_option = I2C_FIRST_AND_LAST_FRAME;
+
+#if DEVICE_I2CSLAVE
+    /* I2C master by default */
+    obj_s->slave = 0;
+#endif
+}
+
+/** Configure the I2C frequency
+ *
+ *  @param obj The I2C object
+ *  @param hz  Frequency in Hz
+ */
+void i2c_frequency(i2c_t *obj, int hz)
+{
+    int timeout;
+    struct i2c_s *obj_s = I2C_S(obj);
+
+    /* wait until I2C_FLAG_I2CBSY flag is reset */
+    timeout = BUSY_TIMEOUT;
+    while ((i2c_flag_get(obj_s->i2c, I2C_FLAG_I2CBSY)) && (--timeout != 0));
+
+    /* reset to clear pending flags */
+    i2c_hw_reset(obj);
+
+    /* disable I2C peripheral */
+    i2c_disable(obj_s->i2c);
+
+    /* configure I2C frequence */
+    i2c_clock_config(obj_s->i2c, hz, I2C_DTCY_2);
+
+    /* configure I2C address mode and slave address */
+    i2c_mode_addr_config(obj_s->i2c, I2C_I2CMODE_ENABLE, I2C_ADDFORMAT_7BITS, 0);
+
+    /* enable I2C peripheral */
+    i2c_enable(obj_s->i2c);
+}
+
+/** Reset I2C peripheral. TODO: The action here. Most of the implementation sends stop()
+ *

```

Comment:
Anything pending here?

Reply:
The comment is copied from \mbed-os\hal\i2c_api.h line 126 and we refer to it.